### DISCUSSION TITLE
Does the `nodelist=G.nodes()` default argument for `normalized_laplacian_matrix` make sense?

### DISCUSSION THREAD

**Discussion by henrykenlay** [networkx/networkx#4637] [Q&A] [answered]
The [function which returns the normalized laplacian matrix](https://networkx.org/documentation/stable//reference/generated/networkx.linalg.laplacianmatrix.normalized_laplacian_matrix.html) has an argument `nodelist=None`. It the `nodelist` is not specified then it used `nodelist=G.nodes()` as the ordering. Consider the following code snippet: 

```
import networkx as nx
import numpy as np

graph = nx.Graph()
graph.add_nodes_from([0, 1, 2, 3, 4])
graph.add_edges_from([[0, 1], [1, 2], [2, 3], [2, 4], [3, 4]])
node_permutation = {0: 2, 1: 1, 2: 0, 3 : 3, 4 : 4}
graph_perm = nx.relabel_nodes(graph, node_permutation)
print('Degree of node 0:', graph.degree(0), graph_perm.degree(0))

print('Expected:,', -1/np.sqrt(graph.degree(0) * graph.degree(1)), -1/np.sqrt(graph_perm.degree(0) * graph_perm.degree(1)))
print('Result (using graph_perm)', nx.normalized_laplacian_matrix(graph)[0, 1], nx.normalized_laplacian_matrix(graph_perm)[0, 1])
print('Result (using .nodes())', nx.normalized_laplacian_matrix(graph)[0, 1], nx.normalized_laplacian_matrix(graph, nodelist=graph_perm.nodes())[0, 1])
print('Result (using sorted .nodes())', nx.normalized_laplacian_matrix(graph)[0, 1], nx.normalized_laplacian_matrix(graph_perm, nodelist=np.arange(5))[0, 1])
```

The first 3 lines create a small graph. Lines 4 and 5 make an isomorphic graph from a permutation that swaps nodes 0 and 2. Line 6 verifies the 5th line seems to work as expected by demonstrating that the degree of node 0 has changed. 

The first print statement gives what I expect the (0, 1) entry of the normalised Laplacian matrix will be for `graph` and `graph_perm`. They are different because the graphs have been relabelled, so the degree of the endpoints of the edge {0, 2} are different. 

The second print statement shows that the entry is the same. In fact, the entire normalised Laplacian matrix is the same. Why is this? Its because the nodelist undoes the permutation before constructing the normalised Laplacian matrix. 

To get what I expected there are two solutions (the last two print statements). The first uses `nx.normalized_laplacian_matrix` on the unpermuted graph and passes the permutation as the `nodelist`. 

The second uses `nodelist = np.arange(5)`. Since `G.nodes()` undoes the permutation when using this function could a reasonable default for `nodelist` instead be `nodelist=np.arange(G.number_of_nodes())`?

**Comment by dschult** [comment, marked as answer]
I can imagine that this specific example might work best using `nodelist=np.arange(len(G))`. But many if not most graphs don't have sequential integers for their nodes. Strings are quite common for nodes, and while integers are also quite popular, they often have a missing or skipped integer. More elaborate choices for nodes are sometimes used -- a node can be any hashable object. They don't even need to be sortable.  Think of a graph where the nodes are graphs <head bender alert>.

My personal preference is that anyone expecting to use array representations of a graph should keep a list of nodes (`nodelist`?) object to allow mapping between the nodes and the rows/columns of the array. The `nodelist` object maps from index to node object. To go the other direction, use `node_to_index = {n: i for i, n in enumerate(G)}`.

In your example, in this paradigm you would use `nodelist=list(arange(len(graph_perm)))` after creating the graph.  Others might use `list(G)` or `sorted(G)` or `sorted(G, key=lambda n: G.node[n]["label"])`.

**Reply by henrykenlay** [reply]
Thank you for your answer @dschult, makes perfect sense!
